File: python/scripts/attachment.py
import os
import sys
import json
import mimetypes
import time
from .api import api_post, extract_id


def _post_file(upload_url, form_data, filepath, timeout=30):
    """S3-style multipart upload (separate from Outline API auth)."""
    import httpx
    with open(filepath, "rb") as f:
        files = {"file": (os.path.basename(filepath), f)}
        return httpx.post(upload_url, data=form_data, files=files, timeout=timeout)


def handle_upload(args):
    filepath = args.filepath
    document_id = extract_id(args.document_id, "doc")
    if not os.path.exists(filepath):
        print(json.dumps({"ok": False, "error": f"File not found: {filepath}"}, ensure_ascii=False))
        sys.exit(1)

    size = os.path.getsize(filepath)
    name = os.path.basename(filepath)
    content_type, _ = mimetypes.guess_type(filepath)
    if not content_type:
        content_type = "application/octet-stream"

    payload = {
        "name": name,
        "size": size,
        "contentType": content_type,
        "documentId": document_id,
    }

    res = api_post("attachments.create", payload)
    if not res.get("ok"):
        print(json.dumps(res, indent=2, ensure_ascii=False))
        sys.exit(1)

    upload_url = res["data"]["uploadUrl"]
    form_data = res["data"]["form"]
    attachment_url = res["data"]["attachment"]["url"]

    upload_res = _post_file(upload_url, form_data, filepath)

    if upload_res.status_code in (200, 204):
        print(
            json.dumps(
                {
                    "ok": True,
                    "attachment_url": attachment_url,
                    "markdown": f"![{name}]({attachment_url})",
                },
                indent=2,
                ensure_ascii=False,
            )
        )
    else:
        print(json.dumps({"ok": False, "status": upload_res.status_code, "error": upload_res.text}, ensure_ascii=False))


def upload_image_for_document(filepath, document_id, max_retries=3):
    """Used by upload_dir to upload local images referenced in markdown."""
    import httpx
    if not os.path.exists(filepath):
        return None

    size = os.path.getsize(filepath)
    name = os.path.basename(filepath)
    content_type, _ = mimetypes.guess_type(filepath)
    if not content_type:
        content_type = "application/octet-stream"

    payload = {
        "name": name,
        "size": size,
        "contentType": content_type,
        "documentId": document_id,
    }

    for attempt in range(max_retries):
        try:
            res = api_post("attachments.create", payload)
            if not res.get("ok"):
                print(f"Failed to get upload URL for {filepath}: {res}")
                return None

            upload_url = res["data"]["uploadUrl"]
            form_data = res["data"]["form"]
            attachment_url = res["data"]["attachment"]["url"]

            upload_res = _post_file(upload_url, form_data, filepath, timeout=60)

            if upload_res.status_code in (200, 204):
                return attachment_url
            print(f"Failed to upload {filepath}: {upload_res.status_code} {upload_res.text[:200]}")
            return None

        except httpx.RequestError as e:
            if attempt < max_retries - 1:
                print(f"Connection error uploading {filepath}, retrying ({attempt + 1}/{max_retries})...")
                time.sleep(2)
            else:
                print(f"Failed to upload {filepath} after {max_retries} attempts: {e}")
                return None
    return None


def setup_attachment_parser(subparsers):
    p = subparsers.add_parser("upload", help="向某文档上传附件/图片")
    p.add_argument("--filepath", required=True, help="本地文件路径")
    p.add_argument("--document-id", required=True, help="文档 ID 或 URL")
    p.set_defaults(func=handle_upload)

File: node/lib/attachment.mjs
import fs from "node:fs";
import path from "node:path";
import { apiPost, extractId } from "./api.mjs";
import { printJson } from "./utils.mjs";

function guessMime(filepath) {
  const ext = path.extname(filepath).toLowerCase();
  const map = {
    ".png": "image/png", ".jpg": "image/jpeg", ".jpeg": "image/jpeg",
    ".gif": "image/gif", ".webp": "image/webp", ".svg": "image/svg+xml",
    ".bmp": "image/bmp", ".ico": "image/x-icon",
    ".pdf": "application/pdf", ".md": "text/markdown",
    ".txt": "text/plain", ".json": "application/json",
    ".zip": "application/zip", ".tar": "application/x-tar",
    ".mp4": "video/mp4", ".mp3": "audio/mpeg",
  };
  return map[ext] || "application/octet-stream";
}

async function postFile(uploadUrl, formData, filepath, timeoutMs = 60_000) {
  const buf = await fs.promises.readFile(filepath);
  const contentType = guessMime(filepath);
  const blob = new Blob([buf], { type: contentType });
  const form = new FormData();
  for (const [k, v] of Object.entries(formData)) form.append(k, String(v));
  form.append("file", blob, path.basename(filepath));
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetch(uploadUrl, { method: "POST", body: form, signal: controller.signal });
  } finally {
    clearTimeout(timer);
  }
}

export async function handleUpload(_action, f) {
  if (!f.filepath) { printJson({ ok: false, error: "需要 --filepath" }); process.exit(1); }
  if (!f.documentId) { printJson({ ok: false, error: "需要 --document-id" }); process.exit(1); }
  const filepath = f.filepath;
  const documentId = extractId(f.documentId, "doc");
  if (!fs.existsSync(filepath)) {
    printJson({ ok: false, error: `File not found: ${filepath}` });
    process.exit(1);
  }
  const stat = await fs.promises.stat(filepath);
  const name = path.basename(filepath);
  const payload = { name, size: stat.size, contentType: guessMime(filepath), documentId };
  const res = await apiPost("attachments.create", payload);
  if (!res.ok) { printJson(res); process.exit(1); }
  const { uploadUrl, form, attachment } = res.data;
  const uploadRes = await postFile(uploadUrl, form, filepath);
  if (uploadRes.status === 200 || uploadRes.status === 204) {
    printJson({
      ok: true,
      attachment_url: attachment.url,
      markdown: `![${name}](${attachment.url})`,
    });
  } else {
    printJson({ ok: false, status: uploadRes.status, error: await uploadRes.text() });
  }
}

export async function uploadImageForDocument(filepath, documentId, maxRetries = 3) {
  if (!fs.existsSync(filepath)) return null;
  const stat = await fs.promises.stat(filepath);
  const name = path.basename(filepath);
  const payload = { name, size: stat.size, contentType: guessMime(filepath), documentId };
  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      const res = await apiPost("attachments.create", payload);
      if (!res.ok) {
        console.log(`Failed to get upload URL for ${filepath}: ${JSON.stringify(res)}`);
        return null;
      }
      const { uploadUrl, form, attachment } = res.data;
      const uploadRes = await postFile(uploadUrl, form, filepath);
      if (uploadRes.status === 200 || uploadRes.status === 204) return attachment.url;
      console.log(`Failed to upload ${filepath}: ${uploadRes.status}`);
      return null;
    } catch (e) {
      if (attempt < maxRetries - 1) {
        console.log(`Connection error uploading ${filepath}, retrying (${attempt + 1}/${maxRetries})...`);
        await new Promise((r) => setTimeout(r, 2000));
      } else {
        console.log(`Failed to upload ${filepath} after ${maxRetries} attempts: ${e.message || e}`);
        return null;
      }
    }
  }
  return null;
}
